src/olmo_core/optim/adamw.py

```diff
@@ -150,3 +214,114 @@ class SkipStepAdamWConfig(OptimConfig):
     @classmethod
     def optimizer(cls) -> Type[SkipStepAdamW]:
         return SkipStepAdamW
+
+
+class SkipStepAdamWForeach(SkipStepOptimizer):
+    """
+    A "skip step" version of AdamW that updates parameters using PyTorch
+    multi‐tensor (*foreach*) kernels.  This executes the heavy math in a handful
+    of fused CUDA kernels instead of one kernel per parameter making it much
+    faster and fully traceable by ``torch.compile``.
+    """
+
+    def __init__(
+        self,
+        params,
+        lr: float = 1e-3,
+        betas: Tuple[float, float] = (0.9, 0.999),
+        eps: float = 1e-8,
+        weight_decay: float = 1e-2,
+        rolling_interval_length: int = 128,
+        sigma_factor: int = 6,
+        dtype: Optional[Union[torch.dtype, DType]] = None,
+    ) -> None:
+        assert lr > 0.0
+        assert all([0.0 <= beta <= 1.0 for beta in betas])
+        defaults = dict(lr=lr, betas=betas, eps=eps, weight_decay=weight_decay)
+        super().__init__(
+            params,
+            defaults,
+            rolling_interval_length=rolling_interval_length,
+            sigma_factor=sigma_factor,
+        )
+        if isinstance(dtype, DType):
+            dtype = dtype.as_pt()
+        self.dtype = dtype
+        self._step_skipped: Optional[torch.Tensor] = None
+
+    @property
+    def step_skipped(self) -> torch.Tensor:
+        if self._step_skipped is not None:
+            return self._step_skipped
+        else:
+            return torch.tensor(0.0)
+
+    # ---------------------------------------------------------------------
+    # Optimizer step – foreach implementation
+    # ---------------------------------------------------------------------
+    @torch.no_grad()
+    def step(self, closure=None) -> None:
+        if closure is not None:
+            with torch.enable_grad():
+                closure()
+
+        step_factor = self.get_step_factor()
+        self._step_skipped = 1 - step_factor
+        if float(step_factor) == 0.0:
+            # Entire step skipped – nothing to do.
+            return
```

Comment:
This is the host-device sync we want to avoid.

Reply:
fixed